docs/2020-02-05-Proving-Lemmas.md

```diff
@@ -0,0 +1,301 @@
+Proving lemmas
+==============
+
+Background
+----------
+
+In order to prove rewrite claims, we usually need lemmas describing
+the function symbols we use. As an example, we may have a lemma saying that the
+length of a string is not negative, i.e.
+```
+rule length(S) > 0 => true
+```
+
+The lemmas that I have seen are, usually, given without proofs.
+
+There is an effort by daejunpark and malturki to prove some properties
+for Casper, simulating Coq-style proofs:
+https://github.com/runtimeverification/beacon-chain-verification/tree/static-model-coq-specs-dependencies/casper/k
+
+Summary
+-------
+
+This document attempts to provide some hints about how to prove the lemmas we
+use for various semantics, in a way similar to what daejunpark and malturki
+did, but in a simpler and less powerful way. Hopefully, proofs done this way
+do not require many additional features from the frontend and backend,
+and can be later transformed into full proofs (which may require additional
+features).
+
+These proofs would need the following:
+
+* Ideally, an induction engine. Practically, at least in the beginning:
+  * we would find a set of symbols that generate a sort
+    (e.g. `0` and `\x -> x+1` for natural numbers, initially without any proof)
+  * we would provide an induction hypothesis
+  * we would provide some claims based on that hypothesis
+    (`k` macros would be nice)
+  * the Haskell backend would check the claims.
+  * we could try to become more rigorous in the future.
+* Some tools to direct proofs
+  * Marking rules as concrete / not concrete for a single proof
+  * Providing hints about which splits are allowed
+    * most of these can be done manually, in the specification, but there are
+      other ways.
+
+Example 1
+---------
+
+We have the following definition, taken from the evm repository:
+```
+syntax Bytes ::= "nilBytes"
+               | Int ":" Bytes
+
+syntax Int ::= lengthBytes(Bytes)
+               [function, functional, smtlib(lengthBytes)]
+syntax Int ::= lengthBytes(Bytes, Int)
+               [function, klabel(lengthBytesAux), smtlib(lengthBytesAux)]
+rule lengthBytes(BS) => lengthBytes(BS, 0)
+rule lengthBytes(nilBytes, SIZE) => SIZE
+rule lengthBytes(B : BS, SIZE) => lengthBytes(BS, SIZE +Int 1)
+```
+
+And we would like to prove the following lemmas:
+
+```
+rule lengthBytes(BS, i) ==Int i +Int lengthBytes(BS)
+rule lengthBytes(BS) >=Int 0 => true
+rule lengthBytes(B:BS) ==Int 1 +Int lengthBytes(BS)
+```
+We would prove them by induction on BS. The set of constructors/generators, is,
+of course, `nilBytes` and `:`.
+
+Let’s start with the first claim. We will start by defining a macro:
+```
+P(i, BS) :=  lengthBytes(BS, i) ==Int i +Int lengthBytes(BS)
+```
+
+Next, we will attempt to prove `P(i, nilBytes)` and
```

Comment:
Maybe justify better how induction is used here, by clearly stating the property proved by induction and the goals derived by induction from it.